packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,889 @@
+        // Set it upfront, such that if connection is established (NoDeltaConnection) or rejected (bug in
+        // connectToDeltaStream() implementation - throwing exception vs. returning rejected promise) in
+        // synchronous way, we have this.connectionP setup for all the code to assert correctness of the flow.
+        this.pendingConnection = true;
+
+        let delayMs = InitialReconnectDelayInMs;
+        let connectRepeatCount = 0;
+        const connectStartTime = performance.now();
+        let lastError: any;
+
+        // This loop will keep trying to connect until successful, with a delay between each iteration.
+        while (connection === undefined) {
+            if (this.closed) {
+                throw new Error("Attempting to connect a closed DeltaManager");
+            }
+            connectRepeatCount++;
+
+            try {
+                this.client.mode = requestedMode;
+                connection = await docService.connectToDeltaStream(this.client);
+
+                if (connection.disposed) {
+                    // Nobody observed this connection, so drop it on the floor and retry.
+                    this.logger.sendTelemetryEvent({ eventName: "ReceivedClosedConnection" });
+                    connection = undefined;
+                }
+            } catch (origError) {
+                if (typeof origError === "object" && origError !== null &&
+                    origError?.errorType === DeltaStreamConnectionForbiddenError.errorType) {
+                    connection = new NoDeltaStream();
+                    requestedMode = "read";
+                    break;
+                }
+
+                // Socket.io error when we connect to wrong socket, or hit some multiplexing bug
+                if (!canRetryOnError(origError)) {
+                    const error = normalizeError(origError);
+                    this.props.closeHandler(error);
+                    throw error;
+                }
+
+                // Log error once - we get too many errors in logs when we are offline,
+                // and unfortunately there is no reliable way to detect that.
+                if (connectRepeatCount === 1) {
+                    logNetworkFailure(
+                        this.logger,
+                        {
+                            delay: delayMs, // milliseconds
+                            eventName: "DeltaConnectionFailureToConnect",
+                            duration: TelemetryLogger.formatTick(performance.now() - connectStartTime),
+                        },
+                        origError);
+                }
+
+                lastError = origError;
+
+                const retryDelayFromError = getRetryDelayFromError(origError);
+                delayMs = retryDelayFromError ?? Math.min(delayMs * 2, MaxReconnectDelayInMs);
+
+                if (retryDelayFromError !== undefined) {
+                    this.props.reconnectionDelayHandler(retryDelayFromError, origError);
```

Comment:
Can/should the `ConnectionManager` emit the delay info directly, rather than calling back to the `DeltaManager` to do it?  Delays in reconnection seem like a job a `ConnectionManager` should report on.

Reply:
Probably similar for `pongHandler`

Reply:
The way it's structured right now, ConnectionManager assumes 1:1 usage pattern (not 1:many), so events is not the right way to do notifications. If we change design, then events would make more sense.